File: 8_StringToInteger/python/StringToInteger.py
class Solution:
    def myAtoi(self, str: str) -> int:
        if len(str) == 0:
            return 0
        
        while str[0] == ' ':
            if len(str) == 1:
                return 0
            str = str[1:]

        
        sign = 1
        if str[0]=='+':
            sign = 1
            str = str[1:]
            
        elif str[0]=='-':
            sign = -1
            str = str[1:]
            
        elif str[0]>='0' and str[0]<='9':
            sigh = 1
            
        else:
            return 0
        
        if len(str) == 0:
            return 0
        
        res = 0
        int_max = pow(2,31)-1
        int_min = -pow(2,31)
        while str[0]>='0' and str[0]<='9':
            res *= 10
            res += int(str[0])
            if res*sign>int_max:
                return int_max
            if res*sign<int_min:
                return int_min
            if len(str)>1:
                
                str = str[1:]
            else:
                break
        return res*sign

File: 8_StringToInteger/c++/StringToInteger.cpp
class Solution {
public:
    int myAtoi(string str) {
        if(str.size() == 0){
            return 0;
        }
        while(str[0] == ' '){
            if(str.size()==1){
                return 0;
            }
            str = str.substr(1,str.size()-1);
        }
        
        int sign;
        if(str[0] == '+'){
            if(str.size()==1){
                return 0;
            }
            str = str.substr(1,str.size()-1);
            sign = -1;
        }
        else if(str[0] == '-'){
            if(str.size()==1){
                return 0;
            }
            str = str.substr(1,str.size()-1);
            sign = 1;
        }
        else if(str[0]>='0' && str[0]<='9'){
            sign = -1;
        }
        else{
            return 0;
        }
        
        if(str[0]<'0' || str[0]>'9'){
            return 0;
        }
        
        int res = -(str[0]-'0');
        if(str.size()==1){
            return res*sign;
        }
        else{
            str = str.substr(1,str.size()-1);
        }
        
        while(str[0]>='0' && str[0]<='9'){
            if(res<(INT_MIN+(str[0]-'0'))/10){
                if(sign == -1){
                    return INT_MAX;
                }
                else{
                    return INT_MIN;
                }
            }

            res = 10 * res - (str[0]-'0');

            if(str.size() == 1){
                break;
            }
            else{
                str = str.substr(1,str.size()-1);
            }
        }
        if(res == INT_MIN && sign == -1){
            return INT_MAX;
        }
        return res*sign;
    }
};